### PR TITLE
[IMP]optimized generic oauth authentication

### PR DESCRIPTION
**Pull request by TimeBye**
**What this PR does / why we need it:** Adding user attributes path for Generic OAuth Authentication

### CODE REVIEW

File: pkg/login/social/generic_oauth.go [hunk 1]
```diff
@@ -189,7 +191,12 @@ func (s *SocialGenericOAuth) UserInfo(client *http.Client, token *oauth2.Token)
 			return nil, fmt.Errorf("Error getting user info: %s", err)
 		}
 
-		err = json.Unmarshal(response.Body, &data)
+		if len(s.userInfoRootPath) != 0 {
+			user := gjson.Get(string(response.Body), s.userInfoRootPath).String()
```
Comment (redbaron) [inline pkg/login/social/generic_oauth.go]: pasrsing, serializaing and then parsing again looks redundant, can't gjson do unmarshalling straight into &data?

Comment (TimeBye) [reply pkg/login/social/generic_oauth.go]: Gjson didn't provide such a method, and I also felt that it was not elegant enough.   😅

**Comment by TimeBye** [comment]
@bergquist  Hello, Could you help me with this problem? This feature really needs to be implemented.
This feature need to be implemented because the user attributes obtained from spring-security-oauth2 is not in the root path.
https://docs.spring.io/spring-security/site/docs/5.2.0.BUILD-SNAPSHOT/api/org/springframework/security/core/userdetails/User.html#User-java.lang.String-java.lang.String-boolean-boolean-boolean-boolean-java.util.Collection-

I refer to the implementation of Gitlab Generic OAuth, in fact, it does the same.
https://gitlab.com/satorix/omniauth-oauth2-generic/tree/master#configuration-options

**Comment by marefr** [comment]
@TimeBye Thank you for your contribution. Seems like this PR solves the same problem as existing PR (#14683). We think that #14683 solves the problem in a better way and something that we can build upon further in the Grafana code base. I'm closing this, but please let me know if I misunderstood and this PR solves some other or additional problems and I'll reopen this.

**Pull request closed without merging**